File: Days 061 - 070/Day 68/PairsOfBishopAttacks.cpp
#include <cmath>
#include <iostream>
#include <type_traits>
#include <vector>

struct Bishop
{
	const int x;
	const int y;

	Bishop(int x, int y) noexcept
		: x(x), y(y)
	{ }
};

template <unsigned int M, typename = std::enable_if_t<(M > 0)>>
unsigned int GetBishopAttackPairCount(const std::vector<Bishop>& bishopLocations) noexcept
{
	unsigned int pairCount = 0;
	auto bishopStack = bishopLocations;

	while (bishopStack.size() > 1)
	{
		const Bishop currentBishop = bishopStack.back();

		if (currentBishop.x < M && currentBishop.x >= 0 &&
			currentBishop.y < M && currentBishop.y >= 0)
		{
			for (auto it = bishopStack.crbegin() + 1; it != bishopStack.crend(); it++)
			{
				const Bishop otherBishop = *it;

				if (otherBishop.x >= M || otherBishop.x < 0 ||
					otherBishop.y >= M || otherBishop.y < 0)
				{
					continue;
				}
				
				const unsigned int xDifference = std::abs(currentBishop.x - otherBishop.x);
				const unsigned int yDifference = std::abs(currentBishop.y - otherBishop.y);

				if (xDifference == yDifference)
				{
					pairCount++;
				}
			}
		}

		bishopStack.pop_back();
	}

	return pairCount;
}

int main(int argc, char* argv[])
{
	const std::vector<Bishop> bishops
	{
		{ 0, 0 },
		{ 1, 2 },
		{ 2, 2 },
		{ 4, 0 }
	};

	std::cout << GetBishopAttackPairCount<5>(bishops) << "\n";

	std::cin.get();

	return 0;
}

File: Days 061 - 070/Day 68/PairsOfBishopAttacks.cs
using System;
using System.Collections.Generic;

namespace DailyCodingProblem
{
	internal struct Bishop
	{
		public int X { get; }
		public int Y { get; }

		public Bishop(int x, int y)
		{
			this.X = x;
			this.Y = y;
		}
	}

	internal class Day68
	{
		private static int Main(string[] args)
		{
			List<Bishop> bishops = new List<Bishop>
			{
				new Bishop(0, 0),
				new Bishop(1, 2),
				new Bishop(2, 2),
				new Bishop(4, 0)
			};

			Console.WriteLine(GetBishopAttackPairCount(bishops, 5));

			Console.ReadLine();

			return 0;
		}

		private static int GetBishopAttackPairCount(List<Bishop> bishopLocations, int m)
		{
			if (m <= 0)
			{
				return 0;
			}

			int pairCount = 0;
			var bishopStack = new List<Bishop>(bishopLocations);

			while (bishopStack.Count > 1)
			{
				Bishop currentBishop = bishopStack[bishopStack.Count - 1];

				if (currentBishop.X < m && currentBishop.X >= 0 &&
					currentBishop.Y < m && currentBishop.Y >= 0)
				{
					for (int i = bishopStack.Count - 2; i >= 0; i--)
					{
						Bishop otherBishop = bishopStack[i];

						if (otherBishop.X >= m || otherBishop.X < 0 ||
							otherBishop.Y >= m || otherBishop.Y < 0)
						{
							continue;
						}

						int xDifference = Math.Abs(currentBishop.X - otherBishop.X);
						int yDifference = Math.Abs(currentBishop.Y - otherBishop.Y);

						if (xDifference == yDifference)
						{
							pairCount++;
						}
					}
				}

				bishopStack.RemoveAt(bishopStack.Count - 1);
			}

			return pairCount;
		}
	}
}